File: src/lesswrong.ts
import fetch from 'node-fetch';
import { Config, LessWrongPost } from './types';
import { AIProcessor } from './aiProcessor';

export class LessWrongIntegration {
  private readonly GRAPHQL_ENDPOINT = 'https://www.lesswrong.com/graphql';

  constructor(
    private config: Config,
    private aiProcessor: AIProcessor
  ) {}

  /**
   * Fetch top recent posts from LessWrong
   */
  async fetchTopPosts(): Promise<LessWrongPost[]> {
    const query = `
      query RecentPosts {
        posts(input: {
          terms: {
            view: "magic"
            limit: ${this.config.lesswrongPostCount}
            filter: "frontpage"
          }
        }) {
          results {
            _id
            title
            slug
            baseScore
            postedAt
            user {
              displayName
            }
            contents {
              plaintextDescription
            }
          }
        }
      }
    `;

    try {
      const response = await fetch(this.GRAPHQL_ENDPOINT, {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
        },
        body: JSON.stringify({ query }),
      });

      const data: any = await response.json();

      if (data.errors) {
        console.error('GraphQL errors:', data.errors);
        return [];
      }

      const posts = data.data.posts.results;

      return posts.map((post: any) => ({
        id: post._id,
        title: post.title,
        author: post.user?.displayName || 'Unknown',
        url: `https://www.lesswrong.com/posts/${post._id}/${post.slug}`,
        baseScore: post.baseScore,
        postedAt: post.postedAt,
        content: post.contents?.plaintextDescription || '',
      }));
    } catch (error) {
      console.error('Error fetching LessWrong posts:', error);
      return [];
    }
  }

  /**
   * Fetch and summarize top posts
   */
  async fetchAndSummarizePosts(): Promise<string> {
    console.log('Fetching top LessWrong posts...');
    const posts = await this.fetchTopPosts();

    if (posts.length === 0) {
      return 'No posts available.';
    }

    console.log(`Found ${posts.length} posts. Generating summaries...`);

    const summaries: string[] = [];

    for (const post of posts) {
      const summary = await this.aiProcessor.summarizePost(
        post.title,
        post.url,
        post.content
      );

      const formattedPost = `### [${post.title}](${post.url})
**Author:** ${post.author} | **Score:** ${post.baseScore}

${summary}`;

      summaries.push(formattedPost);
    }

    return summaries.join('\n\n');
  }
}

File: notes_fixer/lesswrong.py
"""LessWrong integration for fetching and summarizing posts."""

from typing import List
import requests
from .types import Config, LessWrongPost
from .ai_processor import AIProcessor


class LessWrongIntegration:
    """Handles fetching and processing LessWrong posts."""

    GRAPHQL_ENDPOINT = "https://www.lesswrong.com/graphql"

    def __init__(self, config: Config, ai_processor: AIProcessor):
        self.config = config
        self.ai_processor = ai_processor

    def fetch_top_posts(self) -> List[LessWrongPost]:
        """Fetch top recent posts from LessWrong."""
        query = f"""
            query RecentPosts {{
                posts(input: {{
                    terms: {{
                        view: "magic"
                        limit: {self.config.lesswrong_post_count}
                        filter: "frontpage"
                    }}
                }}) {{
                    results {{
                        _id
                        title
                        slug
                        baseScore
                        postedAt
                        user {{
                            displayName
                        }}
                        contents {{
                            plaintextDescription
                        }}
                    }}
                }}
            }}
        """

        try:
            response = requests.post(
                self.GRAPHQL_ENDPOINT,
                json={"query": query},
                headers={"Content-Type": "application/json"},
            )
            response.raise_for_status()
            data = response.json()

            if "errors" in data:
                print(f"GraphQL errors: {data['errors']}")
                return []

            posts = data["data"]["posts"]["results"]

            return [
                LessWrongPost(
                    id=post["_id"],
                    title=post["title"],
                    author=post.get("user", {}).get("displayName", "Unknown"),
                    url=f"https://www.lesswrong.com/posts/{post['_id']}/{post['slug']}",
                    base_score=post["baseScore"],
                    posted_at=post["postedAt"],
                    content=post.get("contents", {}).get("plaintextDescription", ""),
                )
                for post in posts
            ]
        except Exception as e:
            print(f"Error fetching LessWrong posts: {e}")
            return []

    def fetch_and_summarize_posts(self) -> str:
        """Fetch and summarize top posts."""
        print("Fetching top LessWrong posts...")
        posts = self.fetch_top_posts()

        if not posts:
            return "No posts available."

        print(f"Found {len(posts)} posts. Generating summaries...")

        summaries: List[str] = []

        for post in posts:
            summary = self.ai_processor.summarize_post(
                post.title, post.url, post.content
            )

            formatted_post = f"""### [{post.title}]({post.url})
**Author:** {post.author} | **Score:** {post.base_score}

{summary}"""

            summaries.append(formatted_post)

        return "\n\n".join(summaries)
